src/network/networkd-ndisc.c

```diff
@@ -239,76 +249,64 @@ static int ndisc_router_generate_addresses(Link *link, unsigned prefixlen, uint3
         ORDERED_HASHMAP_FOREACH(j, link->network->ipv6_tokens, i) {
-                bool have_address = false;
-                _cleanup_(address_freep) Address *new_address = NULL;
-
-                r = address_new(&new_address);
-                if (r < 0)
-                        return log_oom();
-
-                *new_address = *address;
+                _cleanup_free_ struct in6_addr *new_address = NULL;
 
                 if (j->address_generation_type == IPV6_TOKEN_ADDRESS_GENERATION_PREFIXSTABLE
-                    && memcmp(&j->prefix, &addr, FAMILY_ADDRESS_SIZE(address->family)) == 0) {
+                    && IN6_ARE_ADDR_EQUAL(&j->prefix, address)) {
                         /* While this loop uses dad_counter and a retry limit as specified in RFC 7217, the loop
                            does not actually attempt Duplicate Address Detection; the counter will be incremented
                            only when the address generation algorithm produces an invalid address, and the loop
                            may exit with an address which ends up being unusable due to duplication on the link.
                         */
                         for (; j->dad_counter < DAD_CONFLICTS_IDGEN_RETRIES_RFC7217; j->dad_counter++) {
-                                r = make_stableprivate_address(link, &j->prefix, prefixlen, j->dad_counter, &new_address->in_addr.in6);
+                                r = make_stableprivate_address(link, &j->prefix, prefixlen, j->dad_counter, &new_address);
                                 if (r < 0)
+                                        return r;
+                                if (r > 0)
                                         break;
-
-                                if (stableprivate_address_is_valid(&new_address->in_addr.in6)) {
-                                        have_address = true;
-                                        break;
-                                }
                         }
                 } else if (j->address_generation_type == IPV6_TOKEN_ADDRESS_GENERATION_STATIC) {
-                        memcpy(new_address->in_addr.in6.s6_addr + 8, j->prefix.s6_addr + 8, 8);
-                        have_address = true;
-                }
+                        new_address = new(struct in6_addr, 1);
+                        if (!new_address)
+                                return log_oom();
 
-                if (have_address) {
-                        new_address->prefixlen = prefixlen;
-                        new_address->flags = IFA_F_NOPREFIXROUTE|IFA_F_MANAGETEMPADDR;
-                        new_address->cinfo.ifa_prefered = lifetime_preferred;
+                        memcpy(new_address->s6_addr, address->s6_addr, 8);
+                        memcpy(new_address->s6_addr + 8, j->prefix.s6_addr + 8, 8);
+                }
 
+                if (new_address) {
                         r = set_put(addresses, new_address);
-                        if (r < 0)
```

Comment:
`set_put` returns 0 when the item is already in the set, and 1 when it is not. So the check for -EEXIST is not doing the right thing.

Reply:
If I understand correctly...

> `set_put` returns 0 when the item is already in the set, and 1 when it is not.

Right.

> So the check for -EEXIST is not doing the right thing.

But `set_put()` returns `-EEXIST` if the set contains an element whose hash value is equivalent to the new one. So, I think in this case checking `-EEXIST` is correct. No?

Reply:
From the log in `systemd-networkd-tests.py NetworkdRATests.test_ipv6_token_static_multiple`
```
systemd-networkd[862770]: /run/systemd/network/ipv6-prefix-veth-token-static-multiple.network:7: IPv6 token '::1a:2b:3c:4d' is duplicated, ignoring: File exists
systemd-networkd[862770]: /run/systemd/network/ipv6-prefix-veth-token-static-multiple.network:8: IPv6 token '::1a:2b:3c:4d' is duplicated, ignoring: File exists
```
So, the check `-EEXIST` works correctly.

Reply:
Ah... `ordered_set_put()` is a wrapper of `hashmap_put()`. But `set_put()` is not... Ugh.

Reply:
So, you are right here, as here we use `set_put()`.